plugins/crd/plugin_impl_crd.go

```diff
@@ -0,0 +1,345 @@
+// Copyright (c) 2018 Cisco and/or its affiliates.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at:
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// //go:generate protoc -I ./model/pod --go_out=plugins=grpc:./model/pod ./model/pod/pod.proto
+
+package netmesh_plugin_crd
+
+import (
+	"fmt"
+	"reflect"
+	"sync"
+	"time"
+
+	apiextv1beta1 "k8s.io/apiextensions-apiserver/pkg/apis/apiextensions/v1beta1"
+	apiextcs "k8s.io/apiextensions-apiserver/pkg/client/clientset/clientset"
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	meta_v1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/client-go/kubernetes"
+	"k8s.io/client-go/rest"
+	"k8s.io/client-go/tools/cache"
+	"k8s.io/client-go/tools/clientcmd"
+	"k8s.io/client-go/util/workqueue"
+
+	"github.com/ligato/cn-infra/config"
+	"github.com/ligato/cn-infra/flavors/local"
+	"github.com/ligato/cn-infra/health/statuscheck"
+	"github.com/ligato/cn-infra/logging"
+	"github.com/ligato/networkservicemesh/pkg/apis/networkservice.mesh/v1"
+	client "github.com/ligato/networkservicemesh/pkg/client/clientset/versioned"
+	factory "github.com/ligato/networkservicemesh/pkg/client/informers/externalversions"
+)
+
+// Plugin watches K8s resources and causes all changes to be reflected in the ETCD
+// data store.
+type Plugin struct {
+	Deps
+
+	stopCh chan struct{}
+	wg     sync.WaitGroup
+
+	k8sClientConfig *rest.Config
+	k8sClientset    *kubernetes.Clientset
+	apiclientset    *apiextcs.Clientset
+	crdClient       client.Interface
+
+	StatusMonitor statuscheck.StatusReader
+}
+
+// Deps defines dependencies of netmesh plugin.
+type Deps struct {
+	local.PluginInfraDeps
+	// Kubeconfig with k8s cluster address and access credentials to use.
+	KubeConfig config.PluginConfig
+}
+
+var (
+	// queue is a queue of resources to be processed. It performs exponential
+	// backoff rate limiting, with a minimum retry period of 5 seconds and a
+	// maximum of 1 minute.
+	queueNS  = workqueue.NewRateLimitingQueue(workqueue.NewItemExponentialFailureRateLimiter(time.Second*5, time.Minute))
```

Comment:
Note I'm creating one queue per CRD because I couldn't figure out a way to demultiplex CRDs in a single queue with the APIs provided by the client queueing library. I tried using namespace as a demultiplexer but decided against that.